pallets/utxo/src/script.rs

```diff
@@ -0,0 +1,150 @@
+enum Public {
+    /// Schnorr public key
+    Schnorr(sr25519::Public),
+}
+
+/// Mintlayer script context.
+struct MLContext<'a> {
+    tx_data: &'a [u8],
+}
+
+impl<'a> MLContext<'a> {
+    fn new(tx_data: &'a [u8]) -> Self {
+        MLContext { tx_data }
+    }
+}
+
+impl chainscript::Context for MLContext<'_> {
+    /// Maximum number of bytes pushable to the stack
+    ///
+    /// We do not limit the size of data pushed into the stack in Mintlayer since programmable pool
+    /// binaries may be fairly large. Maximum size is still subject to `MAX_SCRIPT_SIZE`.
+    const MAX_SCRIPT_ELEMENT_SIZE: usize = usize::MAX;
+
+    /// Maximum number of public keys per multisig
+    const MAX_PUBKEYS_PER_MULTISIG: usize = 20;
+
+    /// Maximum script length in bytes
+    ///
+    /// Set it to 100kB for now to allow for mid-size smart contracts to be included in the script.
+    const MAX_SCRIPT_SIZE: usize = 100 * 1024;
+
+    /// Either parsed signature + metadata or unrecognized signature type.
+    type Signature = Signature;
+
+    /// Either a parsed public key or unrecognized pubkey type.
+    type Public = Result<Public, RawSignatureType>;
+
+    /// Extract a signature and sighash.
+    fn parse_signature(&self, sig: &[u8]) -> Option<Self::Signature> {
+        let (&_sighash_byte, sig) = sig.split_last()?;
+        Some(Signature {
+            signature: sig.to_vec(),
+            sighash: (),
+        })
+    }
+
+    /// Extract a pubkey and check it is in the correct format.
+    ///
+    /// Explanation of return values is analogous to `parse_signature`.
+    ///
+    /// The function returns a Rsult wrapped in an Option type.
+    /// * `None` represents a parsing failure, the transaction is rejected.
+    /// * `Some(Err(x))` represents an unrecognized pubkey type with type ID `x`.
+    ///   The pubkey with unknown type ID always succeeds validation. This allows the type to be
+    ///   allocated later for a new signature scheme without introducing a hard fork.
+    /// * `Some(Ok(pk))` is a successfully processed key.
+    fn parse_pubkey(&self, pk: &[u8]) -> Option<Self::Public> {
+        let &key_type = pk.get(0)?;
+        if (key_type as usize) < Public::VARIANT_COUNT {
+            Public::decode_all(&mut &pk[..]).map(Ok).ok()
+        } else {
+            Some(Err(key_type))
+        }
+    }
+
+    /// Verify signature.
+    fn verify_signature(
+        &self,
+        sig: &Self::Signature,
+        pk: &Self::Public,
+        _subscript: &[u8],
+    ) -> bool {
+        let Signature {
+            signature: sig,
+            sighash: _sighash,
+        } = sig;
+        match pk {
+            Ok(Public::Schnorr(pk)) => (&sig[..]).try_into().map_or(false, |sig| {
+                crypto::sr25519_verify(&sr25519::Signature::from_raw(sig), self.tx_data, pk)
+            }),
+            // Unrecognized signature type => accept the signature.
+            Err(_pk) => true,
```

Comment:
I was just wondering, would it pose any problems if I would submit TXs with unrecognized but also invalid signatures and they were accepted and included in the blockchain (would they be included?) and it was only later found out when Mintlayer would adopt this new signature scheme that the blockchain contains TXs with invalid signatures. Can this "attack" actually happen solely because we're returning `true` here and if so, would it be an actual problem?

Reply:
When new signature types are introduced while blockchain is live, we can only start enforcing the new rules starting at some block height in the future in order to not compromise the existing chain. Bitcoin has another layer of protection against the sort of thing you describe. If a transaction looks like using something reserved for future extensions, the node does not forward it. These are so-called "non-standard" transactions. However, if such a transaction ends up in a block, it is deemed valid according to the consensus rules. So I think we should at some point have a way of differentiating between valid, non-standard and invalid transactions/blocks.

Reply:
I was thinking a bit about how to do the softforing updates without reimplementing Substrate. One fairly generic idea could be to use `enum`s where extra variants could be added as the system is developed. For example, in case of Singatures, it could look something like this:

```rust
#[derive(Encode, Decode, Softfork, ...)]
enum Pubkey {
    // Schnorr has been here since the beginning
    Schnorr([u8; 32]),

    // BLS was introduced starting block 1234
    #[softfork(since = 1234)]
    BLS([u8; 32]),
}
```

Then there would be an `Extensible<T>` type that wraps enums like this so that Substrate does not kick out things it is not (yet) able to decode. The `Encode` and `Decode` traits would have to be implemened manually for the `Extensible` type to get desired behaviour. Usage example:
```rust
    let pk: Extensible<Pubkey> = todo!("something automatically decoded by substrate");
    match pk {
        Extensible::Ok(pubkey) => todo!("do something with pubkey of type Pubkey"),
        Extensible::Unrecognized(_) => todo!("handle possible future extension"),
    }
```

Thoughts and/or comments?

Reply:
This might need some thinking about when it comes to the CTs if we go with some ZK stuff as they won't slot nicely in to this architecture but 1) that's years away and 2) that's a me problem

Reply:
I like that idea a lot